app/controllers/DeclarationController.scala

```diff
@@ -16,35 +16,47 @@
 
 package controllers
 
-import javax.inject.Inject
-import play.api.i18n.{I18nSupport, MessagesApi}
-import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import config.FrontendAppConfig
 import connectors.DataCacheConnector
 import controllers.actions._
-import config.FrontendAppConfig
-import models.Mode
-import pages.ConfirmationPage
+import javax.inject.Inject
+import mapper.CaseRequestMapper
+import models.Confirmation.format
+import models._
 import navigation.Navigator
+import pages.ConfirmationPage
+import play.api.i18n.{I18nSupport, MessagesApi}
 import play.api.mvc.{Action, AnyContent}
+import service.CasesService
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
 import views.html.declaration
 
-import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
 
 class DeclarationController @Inject()(
-                                        appConfig: FrontendAppConfig,
-                                        override val messagesApi: MessagesApi,
-                                        dataCacheConnector: DataCacheConnector,
-                                        navigator: Navigator,
-                                        identify: IdentifierAction,
-                                        getData: DataRetrievalAction
-                                      ) extends FrontendController with I18nSupport {
+                                       appConfig: FrontendAppConfig,
+                                       override val messagesApi: MessagesApi,
+                                       dataCacheConnector: DataCacheConnector,
+                                       navigator: Navigator,
+                                       identify: IdentifierAction,
+                                       getData: DataRetrievalAction,
+                                       service: CasesService
+                                     ) extends FrontendController with I18nSupport {
 
   def onPageLoad(mode: Mode): Action[AnyContent] = (identify andThen getData) { implicit request =>
     Ok(declaration(appConfig, mode))
   }
 
   def onSubmit(mode: Mode): Action[AnyContent] = (identify andThen getData).async { implicit request =>
-    Future.successful(Redirect(navigator.nextPage(ConfirmationPage, mode)(request.userAnswers.get)))
+    val answers: UserAnswers = request.userAnswers.get
+
+    service
+      .createCase(new CaseRequestMapper().map(answers))
+      .map(c => answers.set(ConfirmationPage, Confirmation(c.reference)))
+      .flatMap(updatedAnswers => {
+        dataCacheConnector.save(updatedAnswers.cacheMap)
+          .map(_ => Redirect(navigator.nextPage(ConfirmationPage, mode)(updatedAnswers)))
+      })
```

Comment:
I have a feeling @gokyo wont approve of this... :p

Reply:
Read here:
https://stackoverflow.com/questions/14598990/confused-with-the-for-comprehension-to-flatmap-map-transformation

You can convert this block using a `for-comprehension`
The code will be much more readable

Reply:
As mentioned in the past, the only reason I do it this way is because `map` and `flatMap` is quite readable to me :)

Reply:
Look at the code.
The for-comprehension is very tidy and simple to read

In my opinion map and flatMap are readable, but in this case you have lot of brackets
It is very messy to read